{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import pybedtools\n",
    "import simu\n",
    "import subprocess\n",
    "from pybedtools import BedTool\n",
    "\n",
    "# high-level parameters\n",
    "seed_init = 123\n",
    "n_vars = 10\n",
    "n_background_genes = 100\n",
    "min_exons = 3\n",
    "\n",
    "# simulation parameters\n",
    "n_exons = 2\n",
    "block_range = (30, 200)\n",
    "indel_range = (7, 50)\n",
    "exons_range = (1, 3)\n",
    "\n",
    "# read generation parameters\n",
    "fold = 50\n",
    "frag_size = 300\n",
    "frag_sd = 20\n",
    "read_len = 150\n",
    "\n",
    "# references/output files\n",
    "art_illumina = '/Users/marek.cmero/apps/art_bin_MountRainier/ART_ILLUMINA'\n",
    "genome_fasta = '/Users/marek.cmero/reference/fastas/Homo_sapiens.GRCh38.dna.primary_assembly.fa'\n",
    "gtf_ref = '/Users/marek.cmero/reference/gtf/chess_mini_chr12_ref.gtf'\n",
    "junc_ref_file = '/Users/marek.cmero/reference/gtf/chess_mini_chr12_ref.info'\n",
    "out_prefix = '/Users/marek.cmero/Desktop/output/allvars'\n",
    "control_fasta = '%s-control.fasta' % out_prefix\n",
    "case_fasta = '%s-case.fasta' % out_prefix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# build GTF reference\n",
    "gr = BedTool(gtf_ref)\n",
    "junc_ref = simu.build_junc_ref(junc_ref_file)\n",
    "\n",
    "# make gene start/end reference\n",
    "gene_trees = simu.get_gene_features(gr)\n",
    "\n",
    "# get exons for records that have a gene name\n",
    "all_exons = gr.filter(lambda x: x[2] == 'exon').saveas()\n",
    "all_genes = np.unique([simu.get_gene_name(ex) for ex in all_exons if simu.get_gene_name(ex)!=''])\n",
    "\n",
    "# get valid txs\n",
    "valid_txs, valid_genes = simu.get_valid_txs(all_exons, min_exons)\n",
    "valid_txs = np.unique([tx for tx, gn in valid_txs])\n",
    "available_genes = [gene for gene in all_genes if gene in valid_genes]\n",
    "\n",
    "# params dict\n",
    "params = {'n_exons': n_exons,\n",
    "          'ins_range': indel_range,\n",
    "          'block_range': block_range,\n",
    "          'out_prefix': out_prefix}\n",
    "\n",
    "# cleanup and make outdir\n",
    "if os.path.exists(control_fasta):\n",
    "    os.remove(control_fasta)\n",
    "if os.path.exists(case_fasta):\n",
    "    os.remove(case_fasta)\n",
    "outdir = '/'.join(out_prefix.split('/')[:-1])\n",
    "subprocess.call(['mkdir', '-p', outdir])\n",
    "\n",
    "np.random.seed(seed_init)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Fusions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Generating canonical fusion...\n",
      "Generating canonical fusion...\n",
      "Generating canonical fusion...\n",
      "Generating canonical fusion...\n",
      "Generating canonical fusion...\n",
      "Generating canonical fusion...\n",
      "Generating canonical fusion...\n",
      "Generating canonical fusion...\n",
      "Generating canonical fusion...\n",
      "Generating canonical fusion...\n",
      "Generating EE fusion...\n",
      "Generating EE fusion...\n",
      "Generating EE fusion...\n",
      "Generating EE fusion...\n",
      "Generating EE fusion...\n",
      "Generating EE fusion...\n",
      "Generating EE fusion...\n",
      "Generating EE fusion...\n",
      "Generating EE fusion...\n",
      "Generating EE fusion...\n",
      "Generating NE fusion...\n",
      "Generating NE fusion...\n",
      "Generating NE fusion...\n",
      "Generating NE fusion...\n",
      "Generating NE fusion...\n",
      "Generating NE fusion...\n",
      "Generating NE fusion...\n",
      "Generating NE fusion...\n",
      "Generating NE fusion...\n",
      "Generating NE fusion...\n",
      "Generating INS fusion...\n",
      "Generating INS fusion...\n",
      "Generating INS fusion...\n",
      "Generating INS fusion...\n",
      "Generating INS fusion...\n",
      "Generating INS fusion...\n",
      "Generating INS fusion...\n",
      "Generating INS fusion...\n",
      "Generating INS fusion...\n",
      "Generating INS fusion...\n",
      "Generating unpartnered fusion...\n",
      "Generating unpartnered fusion...\n",
      "Generating unpartnered fusion...\n",
      "Generating unpartnered fusion...\n",
      "Generating unpartnered fusion...\n",
      "Generating unpartnered fusion...\n",
      "Generating unpartnered fusion...\n",
      "Generating unpartnered fusion...\n",
      "Generating unpartnered fusion...\n",
      "Generating unpartnered fusion...\n",
      "CPU times: user 10min 59s, sys: 5.87 s, total: 11min 4s\n",
      "Wall time: 11min 17s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "fuscols = ['loc1', 'gene1', 'tx1', 'insert', 'loc2', 'gene2', 'tx2', 'fusion_type']\n",
    "fus_genes, available_genes = simu.pick_genes(n_vars * 5, available_genes)\n",
    "partner_genes, available_genes = simu.pick_genes(n_vars * 4, available_genes)\n",
    "partner_genes = list(partner_genes) + [''] * n_vars\n",
    "add = [''] * n_vars \\\n",
    "    + ['EE'] * n_vars \\\n",
    "    + ['NE'] * n_vars \\\n",
    "    + ['INS'] * n_vars \\\n",
    "    + [''] * n_vars\n",
    "\n",
    "fusions = pd.DataFrame({'gene1': fus_genes,\n",
    "                        'gene2': partner_genes,\n",
    "                        'add': add})\n",
    "fus_info = []\n",
    "\n",
    "for idx, row in fusions.iterrows():\n",
    "    vartype = 'canonical' if row['add'] == '' else row['add']\n",
    "    vartype = 'unpartnered' if row['gene2'] == '' else vartype\n",
    "    \n",
    "    print('Generating %s fusion...' % vartype)\n",
    "    gene1, gene2 = row['gene1'], row['gene2'] if row['gene2'] != '' else None\n",
    "    tx1 = simu.get_transcripts(gene1, all_exons, valid_txs=valid_txs)[0]\n",
    "    tx2 = simu.get_transcripts(gene2, all_exons, valid_txs=valid_txs)[0] if gene2 else None\n",
    "\n",
    "    add = row['add'] if row['add'] != '' else None\n",
    "    fus_parts = simu.write_fusion((tx1, tx2), (gene1, gene2), all_exons,\n",
    "                                  genome_fasta, params, gene_trees, add=add)\n",
    "    fus_info.append(fus_parts)\n",
    "\n",
    "fus_info = pd.DataFrame(fus_info, columns=fuscols)\n",
    "fus_info.to_csv('%s_fusions_simulated.tsv' % out_prefix, index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## TSVs and splice variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Generating DEL TSV...\n",
      "Generating DEL TSV...\n",
      "Generating DEL TSV...\n",
      "Generating DEL TSV...\n",
      "Generating DEL TSV...\n",
      "Generating DEL TSV...\n",
      "Generating DEL TSV...\n",
      "Generating DEL TSV...\n",
      "Generating DEL TSV...\n",
      "Generating DEL TSV...\n",
      "Generating INS TSV...\n",
      "Generating INS TSV...\n",
      "Generating INS TSV...\n",
      "Generating INS TSV...\n",
      "Generating INS TSV...\n",
      "Generating INS TSV...\n",
      "Generating INS TSV...\n",
      "Generating INS TSV...\n",
      "Generating INS TSV...\n",
      "Generating INS TSV...\n",
      "Generating ITD TSV...\n",
      "Generating ITD TSV...\n",
      "Generating ITD TSV...\n",
      "Generating ITD TSV...\n",
      "Generating ITD TSV...\n",
      "Generating ITD TSV...\n",
      "Generating ITD TSV...\n",
      "Generating ITD TSV...\n",
      "Generating ITD TSV...\n",
      "Generating ITD TSV...\n",
      "Generating PTD TSV...\n",
      "Generating PTD TSV...\n",
      "Generating PTD TSV...\n",
      "Generating PTD TSV...\n",
      "Generating PTD TSV...\n",
      "Generating PTD TSV...\n",
      "Generating PTD TSV...\n",
      "Generating PTD TSV...\n",
      "Generating PTD TSV...\n",
      "Generating PTD TSV...\n",
      "Generating INV TSV...\n",
      "Generating INV TSV...\n",
      "Generating INV TSV...\n",
      "Generating INV TSV...\n",
      "Generating INV TSV...\n",
      "Generating INV TSV...\n",
      "Generating INV TSV...\n",
      "Generating INV TSV...\n",
      "Generating INV TSV...\n",
      "Generating INV TSV...\n",
      "Generating EE TSV...\n",
      "Generating EE TSV...\n",
      "Generating EE TSV...\n",
      "Generating EE TSV...\n",
      "Generating EE TSV...\n",
      "Generating EE TSV...\n",
      "Generating EE TSV...\n",
      "Generating EE TSV...\n",
      "Generating EE TSV...\n",
      "Generating EE TSV...\n",
      "Generating NE TSV...\n",
      "Generating NE TSV...\n",
      "Generating NE TSV...\n",
      "Generating NE TSV...\n",
      "Generating NE TSV...\n",
      "Generating NE TSV...\n",
      "Generating NE TSV...\n",
      "Generating NE TSV...\n",
      "Generating NE TSV...\n",
      "Generating NE TSV...\n",
      "Generating RI TSV...\n",
      "Generating RI TSV...\n",
      "Generating RI TSV...\n",
      "Generating RI TSV...\n",
      "Generating RI TSV...\n",
      "Generating RI TSV...\n",
      "Generating RI TSV...\n",
      "Generating RI TSV...\n",
      "Generating RI TSV...\n",
      "Generating RI TSV...\n",
      "Generating NEJ TSV...\n",
      "Generating NEJ TSV...\n",
      "Generating NEJ TSV...\n",
      "Generating NEJ TSV...\n",
      "Generating NEJ TSV...\n",
      "Generating NEJ TSV...\n",
      "Generating NEJ TSV...\n",
      "Generating NEJ TSV...\n",
      "Generating NEJ TSV...\n",
      "Generating NEJ TSV...\n",
      "Generating US TSV...\n",
      "Generating US TSV...\n",
      "Generating US TSV...\n",
      "Generating US TSV...\n",
      "Generating US TSV...\n",
      "Generating US TSV...\n",
      "Generating US TSV...\n",
      "Generating US TSV...\n",
      "Generating US TSV...\n",
      "Generating US TSV...\n",
      "CPU times: user 18min 19s, sys: 9.24 s, total: 18min 29s\n",
      "Wall time: 19min\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "var_genes, available_genes = simu.pick_genes(n_vars * 10, available_genes)\n",
    "vartypes = ['DEL'] * n_vars \\\n",
    "         + ['INS'] * n_vars \\\n",
    "         + ['ITD'] * n_vars \\\n",
    "         + ['PTD'] * n_vars \\\n",
    "         + ['INV'] * n_vars \\\n",
    "         + ['EE'] * n_vars \\\n",
    "         + ['NE'] * n_vars \\\n",
    "         + ['RI'] * n_vars \\\n",
    "         + ['NEJ'] * n_vars \\\n",
    "         + ['US'] * n_vars\n",
    "varcols = ['loc', 'tx', 'gene', 'size', 'exon', 'vartype']\n",
    "            \n",
    "var_df = pd.DataFrame({'gene': var_genes, 'vartype': vartypes})\n",
    "var_info = []\n",
    "\n",
    "for idx, row in var_df.iterrows():\n",
    "    vartype, gene = row['vartype'], row['gene']\n",
    "    print('Generating %s TSV...' % vartype)\n",
    "    \n",
    "    tx = simu.get_transcripts(gene, all_exons, valid_txs=valid_txs)[0]\n",
    "    if vartype in ['DEL', 'INS', 'ITD']:\n",
    "        varsize, exon = simu.write_indel(tx, all_exons, genome_fasta,\n",
    "                                         indel_range, out_prefix, vartype=vartype)\n",
    "    elif vartype in ['PTD', 'INV']:\n",
    "        varsize, exon = simu.write_large_tsv(tx, all_exons, genome_fasta,\n",
    "                                             out_prefix, exons_range, vartype=vartype)\n",
    "    elif vartype == 'NEJ':\n",
    "        varsize, exon = simu.write_trunc_exons(tx, all_exons, genome_fasta, out_prefix, block_range)\n",
    "    elif vartype == 'US':\n",
    "        tx, loc, stats = simu.write_unannot_splice(gene, all_exons, valid_txs, genome_fasta,\n",
    "                                               out_prefix, junc_ref, gene_trees)\n",
    "        varsize, exon = stats['varsize'], stats['exon']\n",
    "    else:\n",
    "        # NE, EE or RI\n",
    "        tx, loc, stats = simu.write_novel_exon(gene, valid_txs, all_exons,\n",
    "                                       genome_fasta, out_prefix, \n",
    "                                       block_range, gene_trees, vartype=vartype)\n",
    "        while tx == '':\n",
    "            # no valid genes, repick\n",
    "            genes, available_genes = simu.pick_genes(1, available_genes)\n",
    "            gene = genes[0]\n",
    "            tx, loc, stats = simu.write_novel_exon(gene, valid_txs, all_exons,\n",
    "                                                   genome_fasta, out_prefix, \n",
    "                                                   block_range, gene_trees, vartype=vartype)\n",
    "        varsize, exon = stats['varsize'], stats['exon']\n",
    "\n",
    "    chrom = simu.get_tx_chrom(tx, all_exons)\n",
    "    loc = simu.get_gene_loc(chrom, gene_trees, gene)\n",
    "    var_info.append([loc, tx, gene, varsize, exon, vartype])\n",
    "\n",
    "var_info = pd.DataFrame(var_info, columns=varcols)\n",
    "var_info.to_csv('%s_tsvs_splice_simulated.tsv' % out_prefix, index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Write background genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 7min 46s, sys: 3.76 s, total: 7min 50s\n",
      "Wall time: 7min 57s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "bg_set, available_genes = simu.pick_genes(n_background_genes, available_genes)\n",
    "for gene in bg_set:\n",
    "    tx = simu.get_transcripts(gene, all_exons, valid_txs=valid_txs)[0]\n",
    "    exons = all_exons.filter(lambda x: x['transcript_id'] == tx).saveas()\n",
    "    tx_seq, strand = simu.get_seq(exons, genome_fasta)\n",
    "    simu.write_wildtype_sequence(tx_seq, strand, control_fasta, tx)\n",
    "    simu.write_wildtype_sequence(tx_seq, strand, case_fasta, tx)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Generate reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 5.84 ms, sys: 25.1 ms, total: 30.9 ms\n",
      "Wall time: 33.6 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# generate reads with art illumina\n",
    "seeds = np.random.randint(0, 99999, 2)\n",
    "\n",
    "# generate case sample\n",
    "subprocess.call([art_illumina, '-ss', 'HS25', '-i', case_fasta, \n",
    "                 '-p', '-l', str(read_len), '-f', str(fold), '-m', str(frag_size),\n",
    "                 '-s', str(frag_sd), '-rs', str(seeds[0]), '-o', '%s-case_R' % out_prefix])\n",
    "\n",
    "# generate control\n",
    "subprocess.call([art_illumina, '-ss', 'HS25', '-i', control_fasta, \n",
    "             '-p', '-l', str(read_len), '-f', str(fold), '-m', str(frag_size),\n",
    "             '-s', str(frag_sd), '-rs', str(seeds[1]), '-o', '%s-control_R' % out_prefix])\n",
    "\n",
    "# gzip files\n",
    "for sample in ['case', 'control']:\n",
    "    for r in range(2):\n",
    "        outf = open('%s-%s_R%d.fastq.gz' % (out_prefix, sample, (r+1)), 'w')\n",
    "        subprocess.call(['gzip', '-c', '%s-%s_R%d.fq' % (out_prefix, sample, (r+1))], stdout=outf)\n",
    "        outf.close()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
