{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 336,
   "metadata": {},
   "outputs": [],
   "source": [
    "import tempfile\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import collections\n",
    "import re\n",
    "import os\n",
    "import subprocess\n",
    "import pybedtools\n",
    "from intervaltree import Interval, IntervalTree\n",
    "from pybedtools import BedTool\n",
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 337,
   "metadata": {},
   "outputs": [],
   "source": [
    "# simulation parameters\n",
    "SEED_INIT = 123\n",
    "N_VARS = 10\n",
    "N_BACKGROUND_GENES = 100\n",
    "N_EXONS = 2 # fuse first N exons to last N exons\n",
    "MIN_EXONS = 3\n",
    "BLOCK_RANGE = (30, 200)\n",
    "INS_RANGE = (3, 50)\n",
    "FOLD = 50\n",
    "FRAG_SIZE = 300\n",
    "FRAG_SD = 20\n",
    "READ_LEN = 150"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 338,
   "metadata": {},
   "outputs": [],
   "source": [
    "# references\n",
    "ART_ILLUMINA = '/Users/marek.cmero/apps/art_bin_MountRainier/ART_ILLUMINA'\n",
    "GENOME_FASTA = '/Users/marek.cmero/reference/fastas/Homo_sapiens.GRCh38.dna.primary_assembly.fa'\n",
    "GTF_REF = '/Users/marek.cmero/reference/gtf/chess_mini_chr12_ref.gtf'\n",
    "FASTA_OUT = 'cryptic_variants_simu.fasta'\n",
    "CONTROL_FASTA_OUT = 'cryptic_variants_control.fasta'\n",
    "OUTDIR = '/Users/marek.cmero/Desktop/output'\n",
    "\n",
    "# constants\n",
    "BASES = list('GCAT')\n",
    "\n",
    "# cleanup (from older runs)\n",
    "if os.path.exists(FASTA_OUT):\n",
    "    os.remove(FASTA_OUT)\n",
    "if os.path.exists(CONTROL_FASTA_OUT):\n",
    "    os.remove(CONTROL_FASTA_OUT)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 339,
   "metadata": {},
   "outputs": [],
   "source": [
    "# functions\n",
    "def get_gene_name(row):\n",
    "    '''\n",
    "    Prevents KeyError if gene name missing\n",
    "    '''\n",
    "    try:\n",
    "        return row.attrs['gene_name']\n",
    "    except KeyError:\n",
    "        return ''\n",
    "    \n",
    "def get_seq(gr):\n",
    "    '''\n",
    "    Returns a dictionary of exon sequences\n",
    "    and the corresponding strand of the transcript\n",
    "    '''\n",
    "    block_seqs = gr.sequence(fi=GENOME_FASTA, s=True)\n",
    "    block_dict = collections.OrderedDict()\n",
    "    with tempfile.NamedTemporaryFile() as fa_tmp:\n",
    "        fa_tmp.write(bytes(open(block_seqs.seqfn).read(), 'utf-8'))\n",
    "        fa_tmp.flush()\n",
    "\n",
    "        for record in SeqIO.parse(fa_tmp.name, 'fasta'):\n",
    "            block_dict[record.id] = str(record.seq)\n",
    "\n",
    "    strand = re.search('\\(([-+])\\)', next(iter(block_dict.keys())))\n",
    "    assert strand\n",
    "    \n",
    "    return block_dict, strand.group(1)\n",
    "\n",
    "def write_sequence(seq_dict, strand, output_file, name):\n",
    "    '''\n",
    "    Writes sequence dictionary to output file\n",
    "    '''\n",
    "    seq = [seq_dict[ex] for ex in seq_dict.keys()]\n",
    "    seq = seq if strand == '+' else [s for s in reversed(seq)]\n",
    "\n",
    "    with open(output_file, 'a') as fout:\n",
    "        fout.write('>%s\\n' % name)\n",
    "        fout.write(''.join(seq) + '\\n')\n",
    "        \n",
    "def get_chrom_features(chrom, gr):\n",
    "    '''\n",
    "    Get all merged intervals on given chromosome\n",
    "    '''\n",
    "    chrom_features = gr.filter(lambda x: x.chrom == chrom).merge()    \n",
    "    chrom_features = [(g.start, g.end) for g in gr]\n",
    "    \n",
    "    chrom_tree = IntervalTree()\n",
    "    [chrom_tree.addi(s, e) for s, e in chrom_features]\n",
    "\n",
    "    return chrom_tree\n",
    "\n",
    "def get_gene_features(chroms, gr):\n",
    "    '''\n",
    "    Get interval tree for start/ends for \n",
    "    each gene on each chromosome\n",
    "    '''\n",
    "    gn_ref = pd.DataFrame([(g.chrom, g.start, g.end, get_gene_name(g)) for g in gr])\n",
    "    aggregator = {1: lambda x: min(x),\n",
    "                  2: lambda x: max(x)}\n",
    "    gn_ref = gn_ref.groupby([0, 3], as_index=False, sort=False).agg(aggregator)\n",
    "    gn_ref = gn_ref[[0, 1, 2, 3]]\n",
    "    gn_ref.columns = ['chrom', 'start', 'end', 'gene']\n",
    "    gn_ref = gn_ref[gn_ref.gene!='']\n",
    "\n",
    "    ref_trees = {}\n",
    "    for chrom in chroms:\n",
    "        chr_ref = gn_ref[gn_ref.chrom == chrom]\n",
    "        ref_tree = IntervalTree()\n",
    "        for s,e,g in zip(chr_ref['start'].values, chr_ref['end'].values, chr_ref['gene'].values):\n",
    "            ref_tree.addi(s-1, e, g)\n",
    "    ref_trees[chrom] = ref_tree\n",
    "    \n",
    "    return ref_trees\n",
    "\n",
    "def get_exon_seq(ex_list, strand, gr, extended=True):\n",
    "    '''\n",
    "    Extends given exon, or creates a novel downstream\n",
    "    exon with random size and returns its sequence. \n",
    "    If a reference exon exists that already extends\n",
    "    or overlaps the given exon, it will extend or place\n",
    "    the exon past the overlapping exon.\n",
    "    '''\n",
    "    # TODO: make sure exon size doesn't extend to the next exon \n",
    "    block_size = np.random.randint(BLOCK_RANGE[0], BLOCK_RANGE[1])\n",
    "    gap_size = 0 if extended else np.random.randint(BLOCK_RANGE[0], BLOCK_RANGE[1])\n",
    "    \n",
    "    block = ex_list[1] if strand == '+' else ex_list[0]\n",
    "    loc = re.compile('[:\\-\\(\\)]').split(block)\n",
    "    exon_start, exon_end = int(loc[1]), int(loc[2])\n",
    "        \n",
    "    start = exon_end + gap_size if strand == '+' else exon_start - gap_size - block_size\n",
    "    end = start + block_size\n",
    "\n",
    "    ex = get_chrom_features(loc[0], gr.merge())\n",
    "    olap = ex.overlap(int(start), int(end))\n",
    "    if len(olap) > 0:\n",
    "        coords = list(olap)[0]\n",
    "        s, e = int(coords[0]), int(coords[1])\n",
    "        \n",
    "        start = e + gap_size if strand == '+' else s - gap_size - block_size\n",
    "        end = start + block_size\n",
    "    \n",
    "    block_bed = '%s\\t%d\\t%d\\t.\\t1\\t%s' % (loc[0], start, end, strand)\n",
    "    block_bt = BedTool(block_bed, from_string=True)\n",
    "    block_seq, bs = get_seq(block_bt)\n",
    "    ext_seq = ''.join([bs for bs in block_seq.values()])\n",
    "    bloc = ''.join([k for k in block_seq.keys()])\n",
    "    \n",
    "    return ext_seq, bloc\n",
    "\n",
    "def increment_seed(seed, amount=1):\n",
    "    seed += amount\n",
    "    np.random.seed(seed)\n",
    "    return seed\n",
    "\n",
    "def get_random_block(chroms, gene_trees):\n",
    "    '''\n",
    "    Get random block sequence for feature\n",
    "    not overlapping any other genomic features\n",
    "    '''    \n",
    "    block_size = np.random.randint(BLOCK_RANGE[0], BLOCK_RANGE[1])\n",
    "    chrom = np.random.choice(chroms)\n",
    "    chrom_features = gene_trees[chrom]\n",
    "    \n",
    "    chr_range = chr_sizes[('chr%s' % chrom)]\n",
    "    block_start = np.random.randint(chr_range[0], chr_range[1]-block_size)\n",
    "    block_end = block_start + block_size        \n",
    "    \n",
    "    seed = SEED_INIT\n",
    "    seq = 'N'\n",
    "    while 'N' in seq:\n",
    "        # only select sequence if there's no Ns\n",
    "        while chrom_features.overlaps(block_start, block_end):\n",
    "            block_start = np.random.randint(chr_range[0], chr_range[1]-block_size)\n",
    "            block_end = block_start + block_size\n",
    "            \n",
    "            if chrom_features.overlaps(block_start, block_end):\n",
    "                seed = increment_seed(seed)\n",
    "\n",
    "        strand = np.random.choice(['+','-'])\n",
    "        block_bed = '%s\\t%d\\t%d\\t.\\t1\\t%s' % (chrom, block_start, block_end, strand)\n",
    "        block_bt = BedTool(block_bed, from_string=True)\n",
    "        block_seq, bs = get_seq(block_bt)\n",
    "        seq = ''.join([bs for bs in block_seq.values()])\n",
    "        \n",
    "        if 'N' in seq:\n",
    "            seed = increment_seed(seed)\n",
    "\n",
    "    return block_seq, seed\n",
    "\n",
    "def get_random_seq():\n",
    "    '''\n",
    "    Generate random insertion sequence\n",
    "    '''\n",
    "    ins_size = np.random.randint(INS_RANGE[0], INS_RANGE[1])\n",
    "    ins = np.random.choice(BASES, ins_size)\n",
    "    ins = ''.join(ins)\n",
    "    return ins\n",
    "\n",
    "def get_tx_seq(tx, all_exons, front=True, wt_out=True):\n",
    "    '''\n",
    "    Get fusion sequence of given transcript, returning\n",
    "    sequence of first N exons for transcript 1 (front=True)\n",
    "    and N exons for transcript 2 (front=False). By default,\n",
    "    write out wildtype transcript to control reference file.\n",
    "    '''\n",
    "    exons = all_exons.filter(lambda x: x['transcript_id'] == tx).saveas()\n",
    "    tx_seq, s = get_seq(exons)\n",
    "    ex_list = [ex for ex in tx_seq.keys()]\n",
    "    \n",
    "    # pick N 5' exons for tx1 (front) and N 3' exons for tx2 (back)\n",
    "    if front:\n",
    "        ex_list = ex_list[:N_EXONS] if s == '+' else ex_list[-N_EXONS:]\n",
    "    else:\n",
    "        ex_list = ex_list[-N_EXONS:] if s == '+' else ex_list[:N_EXONS]\n",
    "    \n",
    "    # select sequences and reverse order if antisense\n",
    "    seq = [tx_seq[ex] for ex in ex_list]\n",
    "    seq = seq if s == '+' else [s for s in reversed(seq)]  \n",
    "    \n",
    "    if wt_out:\n",
    "        write_sequence(tx_seq, s, CONTROL_FASTA_OUT, tx)\n",
    "    \n",
    "    return seq, s, ex_list\n",
    "\n",
    "def write_fusion(tx1, tx2, all_exons, add=None):\n",
    "    '''\n",
    "    Get left and right sequences of given transcripts\n",
    "    corresponding to the first N exons and last N exons of\n",
    "    transcripts 1 and 2 respectively.\n",
    "    Automatically writes wild type transcript to CONTROL_FASTA_OUT.\n",
    "    '''    \n",
    "    exon_types = ['EE', 'NE', 'INS']\n",
    "    if add and add not in exon_types:\n",
    "        raise ValueError('Invalid exon type to add, expected %s' % exon_types)\n",
    "    \n",
    "    # get sequence for tx1\n",
    "    seq1, strand1, ex1_list = get_tx_seq(tx1, all_exons)\n",
    "    \n",
    "    # add to fusion list\n",
    "    fusion_parts = [tx1]\n",
    "\n",
    "    # extended or novel exon\n",
    "    ext_seq, bloc = '', ''\n",
    "    if add == 'EE':\n",
    "        ext_seq, bloc = get_exon_seq(ex1_list, strand1, all_exons)\n",
    "    elif add == 'NE':\n",
    "        ext_seq, bloc = get_exon_seq(ex1_list, strand1, all_exons, extended=True)\n",
    "    elif add == 'INS':\n",
    "        ext_seq = get_random_seq()\n",
    "        bloc = ext_seq\n",
    "    fusion_parts.append(bloc)\n",
    "\n",
    "    seq2 = ''\n",
    "    if tx2:\n",
    "        # get sequence for tx1\n",
    "        seq2, strand2, ex2_list = get_tx_seq(tx2, all_exons, front=False)\n",
    "\n",
    "        # add to fusion list\n",
    "        fusion_parts.append(tx2)\n",
    "    else:\n",
    "        # unpartnered fusion\n",
    "        block_seq, seed = get_random_block(chroms, gene_trees)\n",
    "        seq2 = [s for s in block_seq.values()]\n",
    "        bloc = ''.join([k for k in block_seq.keys()])\n",
    "        fusion_parts.append(bloc)\n",
    "\n",
    "    seq = ''.join(seq1 + [ext_seq] + seq2)\n",
    "    name = '%s:%s:%s' % (tx1, bloc, tx2) if tx2 else '%s:%s' % (tx1, bloc)\n",
    "    \n",
    "    # write output\n",
    "    with open(FASTA_OUT, 'a') as fout:\n",
    "        fout.write('>%s\\n' % name)\n",
    "        fout.write(seq + '\\n')\n",
    "        \n",
    "    return fusion_parts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 340,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 12.5 s, sys: 235 ms, total: 12.7 s\n",
      "Wall time: 12.9 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# build GTF reference\n",
    "gr = BedTool(GTF_REF) \n",
    "\n",
    "# ensure each transcript in reference has at least N exons\n",
    "all_exons = gr.filter(lambda x: x[2] == 'exon').saveas()\n",
    "all_txs = [(tx['transcript_id'], get_gene_name(tx)) for tx in all_exons]\n",
    "valid_txs = pd.DataFrame(pd.Series(all_txs).value_counts(), columns=['exon_count'])\n",
    "valid_txs = valid_txs[valid_txs.exon_count >= MIN_EXONS]\n",
    "valid_txs = valid_txs.index.values\n",
    "\n",
    "all_genes = np.unique([gene for tx, gene in valid_txs if gene != ''])\n",
    "var_genes = np.empty(0)\n",
    "\n",
    "chr_sizes = pybedtools.chromsizes('hg38')\n",
    "chroms = np.unique([x.chrom for x in all_exons])\n",
    "\n",
    "# make gene start/end reference\n",
    "gene_trees = get_gene_features(chroms, gr)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Generate canonical fusions\n",
    "\n",
    "Select `N_VARS` random genes with `N_VARS` random partners and fuse first `N_EXONS` exons to `N_EXONS` terminal exons."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 341,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Generating NCKAP1L:ALKBH2 fusion...\n",
      "Generating LOC105369676:FAM216A fusion...\n",
      "Generating DIABLO:LOC105370063 fusion...\n",
      "Generating NOP2:SLC6A13 fusion...\n",
      "Generating MRPS35:LOC105369758 fusion...\n",
      "Generating WNT10B:ANKRD33 fusion...\n",
      "Generating ESPL1:DHH fusion...\n",
      "Generating RNFT2:LINC01559 fusion...\n",
      "Generating MYF5:YARS2 fusion...\n",
      "Generating ARL6IP4:NRIP2 fusion...\n",
      "CPU times: user 52.2 s, sys: 601 ms, total: 52.8 s\n",
      "Wall time: 54.5 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# pick fusion genes\n",
    "np.random.seed(SEED_INIT)\n",
    "fus_genes = np.random.choice(all_genes, N_VARS * 2, replace=False)\n",
    "fusions = zip(fus_genes[:N_VARS], fus_genes[N_VARS:])\n",
    "fus_txs = [] # which fusions were used in the fusion; for reference\n",
    "\n",
    "# make fusion genes\n",
    "for gene1, gene2 in fusions:\n",
    "    print('Generating %s:%s fusion...' % (gene1, gene2))\n",
    "\n",
    "    # select first transcript from each gene\n",
    "    tx1 = [tx for tx, gn in valid_txs if gn == gene1][0]\n",
    "    tx2 = [tx for tx, gn in valid_txs if gn == gene2][0]\n",
    "\n",
    "    fus_parts = write_fusion(tx1, tx2, all_exons)\n",
    "    fus_parts.append('(%s:%s)' % (gene1, gene2))\n",
    "    fus_txs.append(fus_parts)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Generate unpartnered fusions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 342,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Generating ETFBKMT unpartnered fusion...\n",
      "Generating LOC105369949 unpartnered fusion...\n",
      "Generating LOC101928416 unpartnered fusion...\n",
      "Generating P3H3 unpartnered fusion...\n",
      "Generating ANO4 unpartnered fusion...\n",
      "Generating ZNF268 unpartnered fusion...\n",
      "Generating COPZ1 unpartnered fusion...\n",
      "Generating KRT4 unpartnered fusion...\n",
      "Generating RITA1 unpartnered fusion...\n",
      "Generating LOC105369989 unpartnered fusion...\n",
      "CPU times: user 24 s, sys: 288 ms, total: 24.3 s\n",
      "Wall time: 25.4 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "var_genes = fus_genes.copy()\n",
    "available_genes = list(set(all_genes).symmetric_difference(var_genes))\n",
    "ufus_genes = np.random.choice(available_genes, N_VARS, replace=False)\n",
    "var_genes = np.concatenate([var_genes, ufus_genes])\n",
    "\n",
    "for idx, gene in enumerate(ufus_genes):\n",
    "    print('Generating %s unpartnered fusion...' % gene)\n",
    "                                          \n",
    "    # select first transcript from each gene\n",
    "    tx = [tx for tx, gn in valid_txs if gn == gene][0]\n",
    "        \n",
    "    fus_parts = write_fusion(tx, None, all_exons)\n",
    "    fus_parts.append('(%s:)' % (gene))\n",
    "    fus_txs.append(fus_parts) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Fusions with extended exon at boundary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 343,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Generating LOC105369795:MED13L fusion with extended exon...\n",
      "Generating MED21:RP11-81H14.2 fusion with extended exon...\n",
      "Generating ZNF740:NOS1 fusion with extended exon...\n",
      "Generating LOC105369743:LOC107984498 fusion with extended exon...\n",
      "Generating SLCO1B3:GYS2 fusion with extended exon...\n",
      "Generating RP13-653N12.1:LINC01498 fusion with extended exon...\n",
      "Generating LOC105369956:LOC100506869 fusion with extended exon...\n",
      "Generating RP11-54A9.1:EIF4B fusion with extended exon...\n",
      "Generating LOC283440:ART4 fusion with extended exon...\n",
      "Generating LOC107984516:WNT1 fusion with extended exon...\n",
      "CPU times: user 55.6 s, sys: 699 ms, total: 56.3 s\n",
      "Wall time: 1min\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# pick fusion genes\n",
    "available_genes = list(set(all_genes).symmetric_difference(var_genes))\n",
    "efus_genes = np.random.choice(available_genes, N_VARS*2, replace=False)\n",
    "var_genes = np.concatenate([var_genes, efus_genes])\n",
    "\n",
    "# make fusion genes\n",
    "fusions = zip(efus_genes[:N_VARS], efus_genes[N_VARS:])\n",
    "for gene1, gene2 in fusions:\n",
    "    print('Generating %s:%s fusion with extended exon...' % (gene1, gene2))\n",
    "\n",
    "    # select first transcript from each gene\n",
    "    tx1 = [tx for tx, gn in valid_txs if gn == gene1][0]\n",
    "    tx2 = [tx for tx, gn in valid_txs if gn == gene2][0]\n",
    "\n",
    "    fus_parts = write_fusion(tx1, tx2, all_exons, add='EE')\n",
    "    fus_parts.append('(%s:%s)' % (gene1, gene2))\n",
    "    fus_txs.append(fus_parts)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Fusions with novel exons at boundary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 344,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Generating NDUFA9:LOC105369706 fusion with novel exon...\n",
      "Generating MUC19:OTOGL fusion with novel exon...\n",
      "Generating RP11-900F13.3:IFFO1 fusion with novel exon...\n",
      "Generating LOC105370026:LOC107984442 fusion with novel exon...\n",
      "Generating TCTN2:CD163 fusion with novel exon...\n",
      "Generating LOC105369891:CORO1C fusion with novel exon...\n",
      "Generating GPR162:LOC105369748 fusion with novel exon...\n",
      "Generating LOC101927267:FGD6 fusion with novel exon...\n",
      "Generating ALDH1L2:LOC105369850 fusion with novel exon...\n",
      "Generating MFSD5:LOC101927058 fusion with novel exon...\n",
      "CPU times: user 55.1 s, sys: 699 ms, total: 55.8 s\n",
      "Wall time: 59.6 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# pick fusion genes\n",
    "available_genes = list(set(all_genes).symmetric_difference(var_genes))\n",
    "nfus_genes = np.random.choice(available_genes, N_VARS*2, replace=False)\n",
    "var_genes = np.concatenate([var_genes, nfus_genes])\n",
    "\n",
    "# make fusion genes\n",
    "fusions = zip(nfus_genes[:N_VARS], nfus_genes[N_VARS:])\n",
    "for gene1, gene2 in fusions:\n",
    "    print('Generating %s:%s fusion with novel exon...' % (gene1, gene2))\n",
    "\n",
    "    # select first transcript from each gene\n",
    "    tx1 = [tx for tx, gn in valid_txs if gn == gene1][0]\n",
    "    tx2 = [tx for tx, gn in valid_txs if gn == gene2][0]\n",
    "\n",
    "    fus_parts = write_fusion(tx1, tx2, all_exons, add='NE')\n",
    "    fus_parts.append('(%s:%s)' % (gene1, gene2))\n",
    "    fus_txs.append(fus_parts)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Fusions with insertions at boundary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 345,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Generating LOC105369930:CSAD fusion with a boundary insertion...\n",
      "Generating GDF11:CS fusion with a boundary insertion...\n",
      "Generating KLRF2:RPLP0 fusion with a boundary insertion...\n",
      "Generating MON2:LACAT8 fusion with a boundary insertion...\n",
      "Generating LOC105369699:FLJ41278 fusion with a boundary insertion...\n",
      "Generating GLIPR1L2:LOC105369974 fusion with a boundary insertion...\n",
      "Generating LOC107984447:MYBPC1 fusion with a boundary insertion...\n",
      "Generating SLC8B1:LOC101928162 fusion with a boundary insertion...\n",
      "Generating SLC2A14:LOC100507195 fusion with a boundary insertion...\n",
      "Generating LOC107984520:CLEC2A fusion with a boundary insertion...\n",
      "CPU times: user 46.4 s, sys: 404 ms, total: 46.8 s\n",
      "Wall time: 48.1 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# pick fusion genes\n",
    "available_genes = list(set(all_genes).symmetric_difference(var_genes))\n",
    "ifus_genes = np.random.choice(available_genes, N_VARS*2, replace=False)\n",
    "var_genes = np.concatenate([var_genes, ifus_genes])\n",
    "\n",
    "# make fusion genes\n",
    "fusions = zip(ifus_genes[:N_VARS], ifus_genes[N_VARS:])\n",
    "for gene1, gene2 in fusions:\n",
    "    print('Generating %s:%s fusion with a boundary insertion...' % (gene1, gene2))\n",
    "\n",
    "    # select first transcript from each gene\n",
    "    tx1 = [tx for tx, gn in valid_txs if gn == gene1][0]\n",
    "    tx2 = [tx for tx, gn in valid_txs if gn == gene2][0]\n",
    "\n",
    "    fus_parts = write_fusion(tx1, tx2, all_exons, add='INS')\n",
    "    fus_parts.append('(%s:%s)' % (gene1, gene2))\n",
    "    fus_txs.append(fus_parts)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Write background genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 346,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 3min 53s, sys: 2.02 s, total: 3min 55s\n",
      "Wall time: 4min 1s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# write background genes\n",
    "available_genes = list(set(all_genes).symmetric_difference(var_genes))\n",
    "bg_set = np.random.choice(available_genes, N_BACKGROUND_GENES)\n",
    "for gene in bg_set:\n",
    "    tx = [tx for tx, gn in valid_txs if gn == gene][0]\n",
    "    exons = all_exons.filter(lambda x: x['transcript_id'] == tx).saveas()\n",
    "    tx_seq, strand = get_seq(exons)\n",
    "    write_sequence(tx_seq, strand, CONTROL_FASTA_OUT, tx)\n",
    "    write_sequence(tx_seq, strand, FASTA_OUT, tx)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Generate reads with Art"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 347,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 3.25 ms, sys: 14.5 ms, total: 17.7 ms\n",
      "Wall time: 9.26 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# generate reads with art illumina\n",
    "seeds = np.random.randint(0, 99999, 2)\n",
    "\n",
    "subprocess.call(['mkdir', '-p', OUTDIR])\n",
    "\n",
    "# generate case sample\n",
    "subprocess.call([ART_ILLUMINA, '-ss', 'HS25', '-i', FASTA_OUT, \n",
    "                 '-p', '-l', str(READ_LEN), '-f', str(FOLD), '-m', str(FRAG_SIZE),\n",
    "                 '-s', str(FRAG_SD), '-rs', str(seeds[0]), '-o', '%s/case_R' % OUTDIR])\n",
    "\n",
    "# generate control\n",
    "subprocess.call([ART_ILLUMINA, '-ss', 'HS25', '-i', CONTROL_FASTA_OUT, \n",
    "             '-p', '-l', str(READ_LEN), '-f', str(FOLD), '-m', str(FRAG_SIZE),\n",
    "             '-s', str(FRAG_SD), '-rs', str(seeds[1]), '-o', '%s/control_R' % OUTDIR])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 348,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 3.61 ms, sys: 18.7 ms, total: 22.4 ms\n",
      "Wall time: 10.5 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "for sample in ['case', 'control']:\n",
    "    for r in range(2):\n",
    "        outf = open('%s/%s_R%d.fastq.gz' % (OUTDIR, sample, (r+1)), 'w')\n",
    "        subprocess.call(['gzip', '-c', '%s/%s_R%d.fq' % (OUTDIR, sample, (r+1))], stdout=outf)\n",
    "        outf.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 349,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHS.11800.1</td>\n",
       "      <td></td>\n",
       "      <td>CHS.12716.1</td>\n",
       "      <td>(NCKAP1L:ALKBH2)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHS.11139.1</td>\n",
       "      <td></td>\n",
       "      <td>CHS.12756.5</td>\n",
       "      <td>(LOC105369676:FAM216A)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHS.13049.14</td>\n",
       "      <td></td>\n",
       "      <td>CHS.13194.5</td>\n",
       "      <td>(DIABLO:LOC105370063)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CHS.10783.2</td>\n",
       "      <td></td>\n",
       "      <td>CHS.10619.8</td>\n",
       "      <td>(NOP2:SLC6A13)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CHS.11278.2</td>\n",
       "      <td></td>\n",
       "      <td>CHS.11546.3</td>\n",
       "      <td>(MRPS35:LOC105369758)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CHS.11540.2</td>\n",
       "      <td></td>\n",
       "      <td>CHS.11645.2</td>\n",
       "      <td>(WNT10B:ANKRD33)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CHS.11726.4</td>\n",
       "      <td></td>\n",
       "      <td>CHS.11551.3</td>\n",
       "      <td>(ESPL1:DHH)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>CHS.12898.3</td>\n",
       "      <td></td>\n",
       "      <td>CHS.11083.1</td>\n",
       "      <td>(RNFT2:LINC01559)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>CHS.12272.1</td>\n",
       "      <td></td>\n",
       "      <td>CHS.11365.5</td>\n",
       "      <td>(MYF5:YARS2)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>CHS.13078.3</td>\n",
       "      <td></td>\n",
       "      <td>CHS.10681.8</td>\n",
       "      <td>(ARL6IP4:NRIP2)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>CHS.11342.7</td>\n",
       "      <td></td>\n",
       "      <td>12:126076800-126076966(+)</td>\n",
       "      <td>(ETFBKMT:)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>CHS.12619.1</td>\n",
       "      <td></td>\n",
       "      <td>12:36930568-36930660(+)</td>\n",
       "      <td>(LOC105369949:)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>CHS.13332.1</td>\n",
       "      <td></td>\n",
       "      <td>12:33210710-33210862(-)</td>\n",
       "      <td>(LOC101928416:)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>CHS.10810.2</td>\n",
       "      <td></td>\n",
       "      <td>12:52892110-52892219(+)</td>\n",
       "      <td>(P3H3:)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>CHS.12553.6</td>\n",
       "      <td></td>\n",
       "      <td>12:91840647-91840694(+)</td>\n",
       "      <td>(ANO4:)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>CHS.13378.11</td>\n",
       "      <td></td>\n",
       "      <td>12:89082688-89082887(+)</td>\n",
       "      <td>(ZNF268:)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>CHS.11792.4</td>\n",
       "      <td></td>\n",
       "      <td>12:103033294-103033344(-)</td>\n",
       "      <td>(COPZ1:)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>CHS.11695.3</td>\n",
       "      <td></td>\n",
       "      <td>12:48994550-48994729(+)</td>\n",
       "      <td>(KRT4:)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>CHS.12806.4</td>\n",
       "      <td></td>\n",
       "      <td>12:52892110-52892213(+)</td>\n",
       "      <td>(RITA1:)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>CHS.12811.1</td>\n",
       "      <td></td>\n",
       "      <td>12:91840647-91840694(+)</td>\n",
       "      <td>(LOC105369989:)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>CHS.12015.1</td>\n",
       "      <td>12:62850893-62850987(+)</td>\n",
       "      <td>CHS.12862.3</td>\n",
       "      <td>(LOC105369795:MED13L)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>CHS.11261.5</td>\n",
       "      <td>12:27026534-27026592(+)</td>\n",
       "      <td>CHS.12110.3</td>\n",
       "      <td>(MED21:RP11-81H14.2)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>CHS.11715.9</td>\n",
       "      <td>12:53182190-53182246(+)</td>\n",
       "      <td>CHS.12909.7</td>\n",
       "      <td>(ZNF740:NOS1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>CHS.11439.1</td>\n",
       "      <td>12:45518932-45519053(-)</td>\n",
       "      <td>CHS.11410.1</td>\n",
       "      <td>(LOC105369743:LOC107984498)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>CHS.11174.1</td>\n",
       "      <td>12:20815822-20816017(+)</td>\n",
       "      <td>CHS.11185.1</td>\n",
       "      <td>(SLCO1B3:GYS2)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>CHS.13223.2</td>\n",
       "      <td>12:128112569-128112704(-)</td>\n",
       "      <td>CHS.12688.1</td>\n",
       "      <td>(RP13-653N12.1:LINC01498)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>CHS.12638.1</td>\n",
       "      <td>12:105442060-105442191(-)</td>\n",
       "      <td>CHS.11996.4</td>\n",
       "      <td>(LOC105369956:LOC100506869)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>CHS.12229.4</td>\n",
       "      <td>12:76292270-76292426(+)</td>\n",
       "      <td>CHS.11706.5</td>\n",
       "      <td>(RP11-54A9.1:EIF4B)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>CHS.10676.1</td>\n",
       "      <td>12:2770207-2770308(-)</td>\n",
       "      <td>CHS.11106.2</td>\n",
       "      <td>(LOC283440:ART4)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>CHS.11206.2</td>\n",
       "      <td>12:22861224-22861296(+)</td>\n",
       "      <td>CHS.11542.1</td>\n",
       "      <td>(LOC107984516:WNT1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>CHS.10727.2</td>\n",
       "      <td>12:4654462-4654550(+)</td>\n",
       "      <td>CHS.11254.1</td>\n",
       "      <td>(NDUFA9:LOC105369706)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>CHS.11404.2</td>\n",
       "      <td>12:40397911-40397949(+)</td>\n",
       "      <td>CHS.12267.1</td>\n",
       "      <td>(MUC19:OTOGL)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>CHS.12331.1</td>\n",
       "      <td>12:88582138-88582230(+)</td>\n",
       "      <td>CHS.10782.14</td>\n",
       "      <td>(RP11-900F13.3:IFFO1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>CHS.12941.1</td>\n",
       "      <td>12:119213741-119213827(+)</td>\n",
       "      <td>CHS.13028.1</td>\n",
       "      <td>(LOC105370026:LOC107984442)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>CHS.13105.5</td>\n",
       "      <td>12:123671614-123671706(+)</td>\n",
       "      <td>CHS.10845.3</td>\n",
       "      <td>(TCTN2:CD163)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>CHS.12357.1</td>\n",
       "      <td>12:89948248-89948338(+)</td>\n",
       "      <td>CHS.12702.9</td>\n",
       "      <td>(LOC105369891:CORO1C)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>CHS.10807.4</td>\n",
       "      <td>12:6826353-6826410(+)</td>\n",
       "      <td>CHS.11486.1</td>\n",
       "      <td>(GPR162:LOC105369748)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>CHS.11561.2</td>\n",
       "      <td>12:49323257-49323373(-)</td>\n",
       "      <td>CHS.12460.2</td>\n",
       "      <td>(LOC101927267:FGD6)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>CHS.12626.7</td>\n",
       "      <td>12:105073681-105073860(-)</td>\n",
       "      <td>CHS.12233.2</td>\n",
       "      <td>(ALDH1L2:LOC105369850)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>CHS.11723.2</td>\n",
       "      <td>12:53252289-53252388(+)</td>\n",
       "      <td>CHS.11421.2</td>\n",
       "      <td>(MFSD5:LOC101927058)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>CHS.12507.1</td>\n",
       "      <td>CGGTGGTTTGAGCG</td>\n",
       "      <td>CHS.11714.18</td>\n",
       "      <td>(LOC105369930:CSAD)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>CHS.11842.1</td>\n",
       "      <td>GCGCAGGCATCGAAAG</td>\n",
       "      <td>CHS.11882.9</td>\n",
       "      <td>(GDF11:CS)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>CHS.10947.1</td>\n",
       "      <td>CTCGAGTTCC</td>\n",
       "      <td>CHS.12960.5</td>\n",
       "      <td>(KLRF2:RPLP0)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>CHS.12010.13</td>\n",
       "      <td>AACTCAGTGATTGAGC</td>\n",
       "      <td>CHS.13277.2</td>\n",
       "      <td>(MON2:LACAT8)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>CHS.11220.2</td>\n",
       "      <td>CCCTCTATCCTCTCCCCGTCGAGCTTGGG</td>\n",
       "      <td>CHS.12046.2</td>\n",
       "      <td>(LOC105369699:FLJ41278)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>CHS.12205.6</td>\n",
       "      <td>CGCCTCACTACTTTATATCGGGTACGGT</td>\n",
       "      <td>CHS.12720.2</td>\n",
       "      <td>(GLIPR1L2:LOC105369974)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>CHS.13175.6</td>\n",
       "      <td>CGGGCTATGTGCCTCTCGGGCACCGCCGCTGCCGCCTACGGTT</td>\n",
       "      <td>CHS.12565.34</td>\n",
       "      <td>(LOC107984447:MYBPC1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>CHS.12810.14</td>\n",
       "      <td>TTTCGCGTCCGTGGAAGTGGCTCCATTTGTTTG</td>\n",
       "      <td>CHS.10990.1</td>\n",
       "      <td>(SLC8B1:LOC101928162)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>CHS.10856.7</td>\n",
       "      <td>ACCGTATACCGGACAA</td>\n",
       "      <td>CHS.12110.1</td>\n",
       "      <td>(SLC2A14:LOC100507195)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>CHS.10862.6</td>\n",
       "      <td>GTTTTTGCAGTCTAGGGCACCGGTAC</td>\n",
       "      <td>CHS.10948.6</td>\n",
       "      <td>(LOC107984520:CLEC2A)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               0                                            1  \\\n",
       "0    CHS.11800.1                                                \n",
       "1    CHS.11139.1                                                \n",
       "2   CHS.13049.14                                                \n",
       "3    CHS.10783.2                                                \n",
       "4    CHS.11278.2                                                \n",
       "5    CHS.11540.2                                                \n",
       "6    CHS.11726.4                                                \n",
       "7    CHS.12898.3                                                \n",
       "8    CHS.12272.1                                                \n",
       "9    CHS.13078.3                                                \n",
       "10   CHS.11342.7                                                \n",
       "11   CHS.12619.1                                                \n",
       "12   CHS.13332.1                                                \n",
       "13   CHS.10810.2                                                \n",
       "14   CHS.12553.6                                                \n",
       "15  CHS.13378.11                                                \n",
       "16   CHS.11792.4                                                \n",
       "17   CHS.11695.3                                                \n",
       "18   CHS.12806.4                                                \n",
       "19   CHS.12811.1                                                \n",
       "20   CHS.12015.1                      12:62850893-62850987(+)   \n",
       "21   CHS.11261.5                      12:27026534-27026592(+)   \n",
       "22   CHS.11715.9                      12:53182190-53182246(+)   \n",
       "23   CHS.11439.1                      12:45518932-45519053(-)   \n",
       "24   CHS.11174.1                      12:20815822-20816017(+)   \n",
       "25   CHS.13223.2                    12:128112569-128112704(-)   \n",
       "26   CHS.12638.1                    12:105442060-105442191(-)   \n",
       "27   CHS.12229.4                      12:76292270-76292426(+)   \n",
       "28   CHS.10676.1                        12:2770207-2770308(-)   \n",
       "29   CHS.11206.2                      12:22861224-22861296(+)   \n",
       "30   CHS.10727.2                        12:4654462-4654550(+)   \n",
       "31   CHS.11404.2                      12:40397911-40397949(+)   \n",
       "32   CHS.12331.1                      12:88582138-88582230(+)   \n",
       "33   CHS.12941.1                    12:119213741-119213827(+)   \n",
       "34   CHS.13105.5                    12:123671614-123671706(+)   \n",
       "35   CHS.12357.1                      12:89948248-89948338(+)   \n",
       "36   CHS.10807.4                        12:6826353-6826410(+)   \n",
       "37   CHS.11561.2                      12:49323257-49323373(-)   \n",
       "38   CHS.12626.7                    12:105073681-105073860(-)   \n",
       "39   CHS.11723.2                      12:53252289-53252388(+)   \n",
       "40   CHS.12507.1                               CGGTGGTTTGAGCG   \n",
       "41   CHS.11842.1                             GCGCAGGCATCGAAAG   \n",
       "42   CHS.10947.1                                   CTCGAGTTCC   \n",
       "43  CHS.12010.13                             AACTCAGTGATTGAGC   \n",
       "44   CHS.11220.2                CCCTCTATCCTCTCCCCGTCGAGCTTGGG   \n",
       "45   CHS.12205.6                 CGCCTCACTACTTTATATCGGGTACGGT   \n",
       "46   CHS.13175.6  CGGGCTATGTGCCTCTCGGGCACCGCCGCTGCCGCCTACGGTT   \n",
       "47  CHS.12810.14            TTTCGCGTCCGTGGAAGTGGCTCCATTTGTTTG   \n",
       "48   CHS.10856.7                             ACCGTATACCGGACAA   \n",
       "49   CHS.10862.6                   GTTTTTGCAGTCTAGGGCACCGGTAC   \n",
       "\n",
       "                            2                            3  \n",
       "0                 CHS.12716.1             (NCKAP1L:ALKBH2)  \n",
       "1                 CHS.12756.5       (LOC105369676:FAM216A)  \n",
       "2                 CHS.13194.5        (DIABLO:LOC105370063)  \n",
       "3                 CHS.10619.8               (NOP2:SLC6A13)  \n",
       "4                 CHS.11546.3        (MRPS35:LOC105369758)  \n",
       "5                 CHS.11645.2             (WNT10B:ANKRD33)  \n",
       "6                 CHS.11551.3                  (ESPL1:DHH)  \n",
       "7                 CHS.11083.1            (RNFT2:LINC01559)  \n",
       "8                 CHS.11365.5                 (MYF5:YARS2)  \n",
       "9                 CHS.10681.8              (ARL6IP4:NRIP2)  \n",
       "10  12:126076800-126076966(+)                   (ETFBKMT:)  \n",
       "11    12:36930568-36930660(+)              (LOC105369949:)  \n",
       "12    12:33210710-33210862(-)              (LOC101928416:)  \n",
       "13    12:52892110-52892219(+)                      (P3H3:)  \n",
       "14    12:91840647-91840694(+)                      (ANO4:)  \n",
       "15    12:89082688-89082887(+)                    (ZNF268:)  \n",
       "16  12:103033294-103033344(-)                     (COPZ1:)  \n",
       "17    12:48994550-48994729(+)                      (KRT4:)  \n",
       "18    12:52892110-52892213(+)                     (RITA1:)  \n",
       "19    12:91840647-91840694(+)              (LOC105369989:)  \n",
       "20                CHS.12862.3        (LOC105369795:MED13L)  \n",
       "21                CHS.12110.3         (MED21:RP11-81H14.2)  \n",
       "22                CHS.12909.7                (ZNF740:NOS1)  \n",
       "23                CHS.11410.1  (LOC105369743:LOC107984498)  \n",
       "24                CHS.11185.1               (SLCO1B3:GYS2)  \n",
       "25                CHS.12688.1    (RP13-653N12.1:LINC01498)  \n",
       "26                CHS.11996.4  (LOC105369956:LOC100506869)  \n",
       "27                CHS.11706.5          (RP11-54A9.1:EIF4B)  \n",
       "28                CHS.11106.2             (LOC283440:ART4)  \n",
       "29                CHS.11542.1          (LOC107984516:WNT1)  \n",
       "30                CHS.11254.1        (NDUFA9:LOC105369706)  \n",
       "31                CHS.12267.1                (MUC19:OTOGL)  \n",
       "32               CHS.10782.14        (RP11-900F13.3:IFFO1)  \n",
       "33                CHS.13028.1  (LOC105370026:LOC107984442)  \n",
       "34                CHS.10845.3                (TCTN2:CD163)  \n",
       "35                CHS.12702.9        (LOC105369891:CORO1C)  \n",
       "36                CHS.11486.1        (GPR162:LOC105369748)  \n",
       "37                CHS.12460.2          (LOC101927267:FGD6)  \n",
       "38                CHS.12233.2       (ALDH1L2:LOC105369850)  \n",
       "39                CHS.11421.2         (MFSD5:LOC101927058)  \n",
       "40               CHS.11714.18          (LOC105369930:CSAD)  \n",
       "41                CHS.11882.9                   (GDF11:CS)  \n",
       "42                CHS.12960.5                (KLRF2:RPLP0)  \n",
       "43                CHS.13277.2                (MON2:LACAT8)  \n",
       "44                CHS.12046.2      (LOC105369699:FLJ41278)  \n",
       "45                CHS.12720.2      (GLIPR1L2:LOC105369974)  \n",
       "46               CHS.12565.34        (LOC107984447:MYBPC1)  \n",
       "47                CHS.10990.1        (SLC8B1:LOC101928162)  \n",
       "48                CHS.12110.1       (SLC2A14:LOC100507195)  \n",
       "49                CHS.10948.6        (LOC107984520:CLEC2A)  "
      ]
     },
     "execution_count": 349,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.DataFrame(fus_txs)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
